oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -0,0 +1,258 @@
+    fn read_invocation(&self) -> Result<Invocation, OakStatus> {
+        let message = self.receiver.receive()?;
+        Invocation::create(&self.runtime, message)
+    }
+}
+
+/// Oak Node implementation for the gRPC client pseudo-Node.
+impl Node for GrpcClientNode {
+    fn start(self: Box<Self>) -> Result<JoinHandle<()>, OakStatus> {
+        let thread_handle = thread::Builder::new()
+            .name(self.to_string())
+            .spawn(move || self.worker_thread())
+            .expect("Failed to spawn thread");
+        Ok(thread_handle)
+    }
+}
+
+/// Sends gRPC requests to an external gRPC service and returns gRPC responses.
+struct GrpcRequestHandler {
+    /// A gRPC client dispatcher that wraps a gRPC connection and encodes/decodes messages via a
+    /// provided codec.
+    dispatcher: tonic::client::Grpc<tonic::transport::channel::Channel>,
+}
+
+impl GrpcRequestHandler {
+    fn new(connection: tonic::transport::channel::Channel) -> Self {
+        Self {
+            dispatcher: tonic::client::Grpc::new(connection),
+        }
+    }
+
+    /// Checks whether the [`GrpcRequestHandler::dispatcher`] is able to accept new requests.
+    async fn ready(&mut self) -> Result<(), OakStatus> {
+        self.dispatcher.ready().await.map_err(|error| {
+            error!("Service was not ready: {}", error);
+            OakStatus::ErrInternal
+        })?;
+        Ok(())
+    }
+
+    /// Sends an unary gRPC request wrapped in [`GrpcResponse`] to an external gRPC service.
+    ///
+    /// [`GrpcResponse`]: oak_abi::proto::oak::encap::GrpcResponse
+    async fn unary_request(&mut self, request: GrpcRequest) -> Result<GrpcResponse, OakStatus> {
+        self.ready().await?;
+        let codec = VecCodec::default();
+        let path = http::uri::PathAndQuery::from_maybe_shared(bytes::Bytes::from(
+            request.method_name.to_string(),
+        ))
+        .map_err(|error| {
+            error!("Invalid URI {}: {}", request.method_name, error);
+            OakStatus::ErrInternal
+        })?;
+
+        match self
+            .dispatcher
+            .unary(tonic::Request::new(request.req_msg), path, codec)
+            .await
+        {
+            Ok(response) => Ok(GrpcResponse {
+                rsp_msg: response.into_inner(),
+                status: Some(from_tonic_status(tonic::Status::ok(""))),
+                last: true,
+            }),
+            Err(status) => Ok(GrpcResponse {
+                rsp_msg: vec![],
+                status: Some(from_tonic_status(status)),
+                last: true,
+            }),
+        }
+    }
+}
+
+/// Converts [`tonic::Status`] to [`oak_abi::proto::google::rpc::Status`].
+fn from_tonic_status(status: tonic::Status) -> oak_abi::proto::google::rpc::Status {
+    oak_abi::proto::google::rpc::Status {
+        code: status.code() as i32,
+        message: status.message().to_string(),
+        details: vec![prost_types::Any {
+            value: status.details().to_vec(),
```

Comment:
Are we sure that we always want to propagate internal status details? Are they guaranteed not to contain sensitive info?

Reply:
These are details that have been sent back by an external gRPC service.
Since they were sent to the client, I think it's ok to reveal them to the client Node.
But I may be wrong.

Reply:
Makes sense I guess, could you add a comment to remind ourselves of this?